usecase/utils.go

```diff
@@ -0,0 +1,140 @@
+package usecase
+
+import (
+	"api/entities"
+	"encoding/csv"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"sync/atomic"
+)
+
+var readerCounter int32
+
+func ReadCsv(filename string) ([][]string, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	reader := csv.NewReader(file)
+	records, err := reader.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	return records, nil
+}
+
+func ParseRecordsToJSON(records [][]string) ([]byte, error) {
+	recordsMap, err := ParseRecordsToMap(records)
+	if err != nil {
+		return nil, err
+	}
+	response, err := json.Marshal(recordsMap)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func ParseRecordsToMap(records [][]string) (map[int]string, error) {
+	recordsMap := make(map[int]string)
+	for _, line := range records {
+		id, err := strconv.Atoi(line[0])
+		if err != nil {
+			return nil, err
+		}
+		recordsMap[id] = line[1]
+	}
+	return recordsMap, nil
+}
+
+func WriteCSVFromResponse(filename string, data *entities.Response) error {
+	csvFile, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer csvFile.Close()
+
+	writer := csv.NewWriter(csvFile)
+	for index, berry := range data.Results {
+		var row []string
+		row = append(row, strconv.Itoa(index))
+		row = append(row, berry.Name)
+		row = append(row, berry.Url)
+		writer.Write(row)
+	}
+	writer.Flush()
+	return nil
+}
+
+func ValidateConcurrentReadParams(readType string, items int, itemsPerWorker int) (int32, error) {
+	var initialCounterPosition int32
+	var err error
+	if readType == "odd" {
+		initialCounterPosition = 1
+	} else if readType == "even" {
+		initialCounterPosition = 0
+	} else {
+		err = errors.New("Invalid type")
+		return 0, err
+	}
+
+	if items == 0 || itemsPerWorker == 0 {
+		err = errors.New("Invalid items")
+		return 0, err
+	}
+	return initialCounterPosition, nil
+}
+
+func ReadRecordsConcurrently(records [][]string, items int, itemsPerWorker int, initialCounterPosition int32) map[string]string {
```

Comment:
This approach makes sense to me, but I have two concerns about it:
1. I think you're depending on always having an equal amount of odd and even IDs in your records, it's very probable that wouldn't be the case with real-world data.
2. You're also depending on your data being ordered, this is less of a concern since it can be sorted beforehand.

Reply:
1. This is true. This approach reads results based on the indexes more than the ids. In this case they coincide, even if they might not with another set of data. However, I built this function so only the indexes I care for reading are processed instead of processing the whole map, this is why I create the number of workers based of records / 2.
2. This is also true, but because I am accessing the indexes they are always ordered.